python/plugins/processing/algs/qgis/Climb.py

```diff
@@ -0,0 +1,260 @@
+
+
+class Climb(QgisAlgorithm):
+
+    INPUT = 'INPUT'
+    OUTPUT = 'OUTPUT'
+
+    TOTALCLIMB = 'TOTALCLIMB'
+    TOTALDESCENT = 'TOTALDESCENT'
+    MINELEVATION = 'MINELEVATION'
+    MAXELEVATION = 'MAXELEVATION'
+    CLIMBATTRIBUTE = 'climb'
+    DESCENTATTRIBUTE = 'descent'
+    MINELEVATTRIBUTE = 'minelev'
+    MAXELEVATTRIBUTE = 'maxelev'
+
+    def name(self):
+        return 'climbalongline'
+
+    def displayName(self):
+        return self.tr('Climb Along Line')
+
+    def group(self):
+        return self.tr('Vector analysis')
+
+    def groupId(self):
+        return 'vectoranalysis'
+
+    def __init__(self):
+        super().__init__()
+
+    def initAlgorithm(self, config=None):
+        self.addParameter(
+            QgsProcessingParameterFeatureSource(
+                self.INPUT,
+                self.tr('Input (line) layer'),
+                [QgsProcessing.TypeVectorLine]
+            )
+        )
+
+        self.addParameter(
+            QgsProcessingParameterFeatureSink(
+                self.OUTPUT,
+                self.tr('Climb layer')
+            )
+        )
+
+        self.addOutput(
+            QgsProcessingOutputNumber(
+                self.TOTALCLIMB,
+                self.tr('Total climb')
+            )
+        )
+
+        self.addOutput(
+            QgsProcessingOutputNumber(
+                self.TOTALDESCENT,
+                self.tr('Total descent')
+            )
+        )
+
+        self.addOutput(
+            QgsProcessingOutputNumber(
+                self.MINELEVATION,
+                self.tr('Minimum elevation')
+            )
+        )
+
+        self.addOutput(
+            QgsProcessingOutputNumber(
+                self.MAXELEVATION,
+                self.tr('Maximum elevation')
+            )
+        )
+
+    def processAlgorithm(self, parameters, context, feedback):
+
+        source = self.parameterAsSource(
+            parameters,
+            self.INPUT,
+            context
+        )
+
+        fcount = source.featureCount()
+        source_fields = source.fields()
+
+        hasZ = QgsWkbTypes.hasZ(source.wkbType())
+
+        if not hasZ:
+            raise QgsProcessingException(self.tr('The layer does not have Z values. If you have a DEM, use the Drape algorithm to extract Z values.'))
+            return {}
+
+        thefields = QgsFields()
+        climbindex = -1
+        descentindex = -1
+        minelevindex = -1
+        maxelevindex = -1
+        fieldnumber = 0
+
+        # Create new fields for climb and descent
+        thefields.append(QgsField(self.CLIMBATTRIBUTE, QVariant.Double))
+        thefields.append(QgsField(self.DESCENTATTRIBUTE, QVariant.Double))
+        thefields.append(QgsField(self.MINELEVATTRIBUTE, QVariant.Double))
+        thefields.append(QgsField(self.MAXELEVATTRIBUTE, QVariant.Double))
+
+        # combine all the vector fields
+        out_fields = QgsProcessingUtils.combineFields(source_fields, thefields)
```

Comment:
```suggestion
        out_fields = QgsProcessingUtils.combineFields(thefields, source_fields)
```

Reply:
This change will make sure that the names of the generated attributes will be known in the output ("colliding" names in the input will be renamed).